src/main/java/io/github/thebusybiscuit/slimefun4/core/machines/MachineProcessor.java

```diff
@@ -0,0 +1,279 @@
+     * We don't need to validate our input here as that is already
+     * covered in our public methods.
+     * 
+     * @param pos
+     *            The {@link BlockPosition} at which our machine is located.
+     * 
+     * @return The current {@link MachineOperation} or null.
+     */
+    @Nullable
+    private T getOperation(@Nonnull BlockPosition pos) {
+        lock.readLock().lock();
+
+        try {
+            return machines.get(pos);
+        } finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * This will end the {@link MachineOperation} at the given {@link Location}.
+     * 
+     * @param loc
+     *            The {@link Location} at which our machine is located.
+     * 
+     * @return Whether the {@link MachineOperation} was successfully ended. This will return false if there was no
+     *         {@link MachineOperation} to begin with.
+     */
+    public boolean endOperation(@Nonnull Location loc) {
+        Validate.notNull(loc, "The location should not be null");
+
+        return endOperation(new BlockPosition(loc));
+    }
+
+    /**
+     * This will end the {@link MachineOperation} at the given {@link Block}.
+     * 
+     * @param b
+     *            The {@link Block} at which our machine is located.
+     * 
+     * @return Whether the {@link MachineOperation} was successfully ended. This will return false if there was no
+     *         {@link MachineOperation} to begin with.
+     */
+    public boolean endOperation(@Nonnull Block b) {
+        Validate.notNull(b, "The Block should not be null");
+
+        return endOperation(new BlockPosition(b));
+    }
+
+    /**
+     * This will end the {@link MachineOperation} at the given {@link BlockPosition}.
+     * We don't need to validate our input here as that is already
+     * covered in our public methods.
+     * 
+     * @param pos
+     *            The {@link BlockPosition} at which our machine is located.
+     * 
+     * @return Whether the {@link MachineOperation} was successfully ended. This will return false if there was no
+     *         {@link MachineOperation} to begin with.
+     */
+    private boolean endOperation(@Nonnull BlockPosition pos) {
+        lock.writeLock().lock();
+
+        try {
+            T operation = machines.remove(pos);
+
+            if (operation != null) {
+                /*
+                 * Only call an event if the operation actually finished.
+                 * If it was ended prematurely (aka aborted), then we don't call any event.
+                 */
+                if (operation.isFinished()) {
+                    Event event = new AsyncMachineOperationFinishEvent(pos, this, operation);
+                    Bukkit.getPluginManager().callEvent(event);
+                }
+
+                return true;
+            } else {
+                return false;
+            }
+        } finally {
+            lock.writeLock().unlock();
```

Comment:
We're keeping this locked way longer than needed. This should be unlocked as soon as remove is done. Other plugin processing is very unpredictable

Reply:
good point